libcudacxx/include/cuda/std/detail/libcxx/include/__complex/traits.h

```diff
@@ -0,0 +1,165 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCUDACXX___COMPLEX_TRAITS_H
+#define _LIBCUDACXX___COMPLEX_TRAITS_H
+
+#ifndef __cuda_std__
+#  include <config>
+#endif // __cuda_std__
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include "../__type_traits/enable_if.h"
+#include "../__type_traits/is_arithmetic.h"
+#include "../__type_traits/is_floating_point.h"
+#include "../__type_traits/void_t.h"
+#include "../__utility/declval.h"
+#include "../cstddef"
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+template <class _Tp>
+struct __is_nvfp16 : false_type
+{};
+
+template <class _Tp>
+struct __is_nvbf16 : false_type
+{};
+
+template <class _Tp>
+struct __is_complex_float
+{
+  static constexpr auto value =
+    _LIBCUDACXX_TRAIT(is_floating_point, _Tp) || __is_nvfp16<_Tp>::value || __is_nvbf16<_Tp>::value;
+};
+
+template <class _Tp>
+struct __is_complex_arithmetic
+{
+  static constexpr auto value = _LIBCUDACXX_TRAIT(is_arithmetic, _Tp) || __is_complex_float<_Tp>::value;
+};
+
+template <class _Tp>
+struct __complex_alignment : integral_constant<size_t, 2 * sizeof(_Tp)>
+{};
+
+# if _LIBCUDACXX_CUDA_ABI_VERSION > 3
+#  define _LIBCUDACXX_COMPLEX_ALIGNAS _ALIGNAS(__complex_alignment<_Tp>::value)
+# else
+#  define _LIBCUDACXX_COMPLEX_ALIGNAS
+# endif
+
+template <class _Dest, class _Source, class = void>
+struct __complex_can_implicitly_construct : false_type
+{};
+
+template <class _Dest, class _Source>
+struct __complex_can_implicitly_construct<_Dest, _Source, __void_t<decltype(_Dest{_CUDA_VSTD::declval<_Source>()})>>
+    : true_type
+{};
+
+template <class _Tp>
+struct __type_to_vector;
+
+template <class _Tp>
+using __type_to_vector_t = typename __type_to_vector<_Tp>::__type;
+
+template <class _Tp, typename = void>
+struct __has_vector_type : false_type
+{};
+
+template <class _Tp>
+struct __has_vector_type<_Tp, __void_t<__type_to_vector_t<_Tp>>> : true_type
+{};
+
+template <class _Tp>
+struct __abcd_results
+{
+  _Tp __ac;
+  _Tp __bd;
+  _Tp __ad;
+  _Tp __bc;
+};
+
+template <class _Tp>
+struct __ab_results
+{
+  _Tp __a;
+  _Tp __b;
+};
+
+template <class _Tp, typename = __enable_if_t<!__has_vector_type<_Tp>::value>>
+_LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX11 __abcd_results<_Tp>
+__complex_calculate_partials(_Tp __a, _Tp __b, _Tp __c, _Tp __d) noexcept
```

Comment:
I don't know how I feel about these functions being in a "traits" header, because at this point it's more than traits. Maybe rename to something like vector_math or something in a similar vein?

Reply:
I believe it should better be named `helper.h`?